app/app_test.go

```diff
@@ -63,6 +69,64 @@ func TestExport(t *testing.T) {
 	assert.Len(t, exportedApp.Validators, 1) // no validators set in default genesis
 }
 
+// TestLegacyMsgAreAminoRegistered checks if all known msg types are registered on the app's amino codec.
+// It doesn't check if they are registered on the module codecs used for signature checking.
+func TestLegacyMsgAreAminoRegistered(t *testing.T) {
+	tApp := NewTestApp()
+
+	lcdc := tApp.LegacyAmino()
+
+	// Use the proto codec as the canonical list of msg types.
+	protoCodec := tApp.AppCodec().(*codec.ProtoCodec)
+	msgProtoNames := protoCodec.InterfaceRegistry().ListImplementations(sdk.MsgInterfaceProtoName)
+
+	for i, msgName := range msgProtoNames {
+		// skip external unregistered msgs
+		if msgName == sdk.MsgTypeURL(&evmtypes.MsgEthereumTx{}) ||
+			msgName == sdk.MsgTypeURL(&vestingtypes.MsgCreatePeriodicVestingAccount{}) {
+			continue
+		}
+
+		// Create an encoded json msg, then unmarshal it to instantiate the msg type.
+		jsonMsg := []byte(fmt.Sprintf(`{"@type": "%s"}`, msgName))
+
+		var msg sdk.Msg
+		err := protoCodec.UnmarshalInterfaceJSON(jsonMsg, &msg)
+		require.NoError(t, err)
+
+		// Only check legacy msgs for amino registration.
+		// Only legacy msg can be signed with amino.
+		_, ok := msg.(legacytx.LegacyMsg)
+		if !ok {
+			continue
+		}
+
+		// Check the msg is registered in amino by checking a repeat registration call panics.
+		panicValue, ok := catchPanic(func() {
```

Comment:
super clever way to check the messages are registered 👍 🥇 

nit: you can probably use [`require.PanicsWithError`](https://pkg.go.dev/github.com/stretchr/testify/require?utm_source=godoc#PanicsWithValue) (or `PanicsWithError`) to test this.

Reply:
thanks, I started with `PanicsWithValue/Error` but the panic returns an error msg containing the go type name of the msg (eg `"types.MsgPlaceBid"`). I couldn't think of a way to reconstruct it to do the comparison.